### ISSUE TITLE
flag `print` in `assert` message

### ISSUE THREAD

**Issue by janosh** [astral-sh/ruff#11974]
i've come across code like this a few times, leading me to believe a lint rule to catch this makes sense

```py
assert 1 == 2, print("probably didn't mean to print this")
>>> AssertionError: None
```

the developer very likely wanted the `print` message to become the error message but of course the error message is `None` instead, `print`'s return value

**Comment by zanieb** [comment]
I thought there was a rule for this but I don't see one. This makes sense to me.

**Comment by denwong47** [comment]
Seeing that this could be a fairly straight forward condition to check against, I would like to have a go at this as Ruff Rule #030.

I wonder if we have to support all variations of `print` calls, such as:
```python
U00A0 = "\u00a0"
assert False, print("Unreachable due to ", condition, f", ask {maintainer} for advice", sep=U00A0)
```

I can potentially do a `FStringValue::concatenated` as:
```python
assert False, "Unreachable due to " f"{U00A0}" f"{condition}" f"{U00A0}" f", ask {maintainer} for advice"
```
But honestly it does not make very readable code.

Any advice would be welcomed.

**Comment by MichaReiser** [comment]
I wonder if it makes sense to phrase this rule more generally. Overall, it is probably incorrect if anything that returns `None` is passed as the assertion message. But implementing this would likely require type inference.

**Comment by denwong47** [comment]
With type inference this rule could become a nice-to-have: since the return type of print already violates the type of the statement, the only thing that this rule will provide is a nicer error message instead of "str is expected but found NoneType".

**Comment by charliermarsh** [comment]
Added by @denwong47 in https://github.com/astral-sh/ruff/pull/11981. Thank you, great contribution!
